src/daemon/daemon_config.cpp

```diff
@@ -132,10 +133,19 @@ std::unique_ptr<const mp::DaemonConfig> mp::DaemonConfigBuilder::build()
             hosts.push_back(image.get());
         }
 
-        vault = platform::make_image_vault(
-            hosts, url_downloader.get(),
-            mp::utils::backend_directory_path(cache_directory, factory->get_backend_directory_name()),
-            mp::utils::backend_directory_path(data_directory, factory->get_backend_directory_name()), days_to_expire);
+        const auto& driver = utils::get_driver_str();
+        if (driver == QStringLiteral("lxd"))
+        {
+            vault = platform::make_image_vault(factory.get(), driver, hosts);
+        }
+        else
+        {
+            vault = std::make_unique<DefaultVMImageVault>(
+                hosts, url_downloader.get(),
+                mp::utils::backend_directory_path(cache_directory, factory->get_backend_directory_name()),
+                mp::utils::backend_directory_path(data_directory, factory->get_backend_directory_name()),
+                days_to_expire);
+        }
```

Comment:
I wonder if the Factory itself could provide the VMImageVault? To avoid the `if (driver…)` bit?

Reply:
@Saviq,

I thought about that, but then we'd have to modify *all* 6 supported platform factories in order for them to provide the `VMImageVault`.  That seemed like a bunch of work just to make the LXD situation a little cleaner.

Reply:
Well, yeah, but the alternative being special-casing the lxd driver in a bunch of places…

I long suggested we should have `Base*` default implementations of things like the VM factory, maybe this is the right time to introduce it?

Reply:
I, too, am a fan of defining a base implementation with each factory able to override any parts of it as it sees fit.  Of course, this goes against the original philosophy of how the various classes define the interface and then the platform itself defines the implementation.  But given that we keep extending this further and further with new backends, etc., I think that doesn't make sense for the cases of code duplication that each implementation shares.

That said, we do have an example in the code where we do break from the original philosophy where we have a base `CommomVMImageHost` class (based off of `VMImageHost`) and `UbuntuVMImageHost` and `CustomVMImageHost` derive from that.

Since this a fairly significant break with (most of) the current code philosophy, I'd like ask @ricab and @luis4a0 their opinions on this. (I know @luis4a0 is on holiday, so his response may be late to the game :grinning: )

Reply:
TBH I don't see this as a departure from the philosophy. We still *only* define the interface first. Any shared implementation comes later, either in the form of a base class, or maybe mixins.

Reply:
Ok, then I'm not sure why we haven't done this sooner :grin: At any rate, then sure, I'm all for defining a shared base class implementation and going from there.  I will see if it makes more sense to do that in a separate PR or do it here and rename the branch, title, etc.

Reply:
This is fixed now.  The first part of this is in #1609 and this PR has been updated to commonize creating the image vault.